### PR TITLE
Fix mbed_enter_sleep function.

### PR DESCRIPTION
**Pull request by pan-**
The previous version of mbed_enter_sleep_function was only working if the
softdevice was enabled. The root cause of this bug is that WFE was called
after sd_softdevice_is_enabled.
A call to sd_softdevice_is_enabled trigger an SVC call and register an
event in the event register. When WFE is called, if an event is
registered, the cpu will not go to sleep.
This means that in the previous version of mbed_enter_sleep, if the
softdevice was not enabled, the CPU never goes to sleep.

The new startegy here is to clear the event register then, if there is
interrupts pending, just return from this function. If no interrupts are
pending, just call WFE.

This PR should be merged with: https://github.com/ARMmbed/core-util/pull/77

### CODE REVIEW

**Comment by pan-** [comment]
@rgrover @0xc0170 could you review this please ?

File: source/sleep.c [hunk 1]
```diff
@@ -37,8 +37,31 @@ void mbed_enter_sleep(sleep_t *obj)
...
+        // interrupt, return from here. If no interrupts are pending, just call
+        // WFE.
+
+        // Set an event and wake up whatsoever, this will clear the event
+        // register from all previous events set (SVC call included)
+        __SEV();
         __WFE();
+
+        // Test if there is an interrupt pending (mask reserved regions)
+        if(SCB->ICSR & (0x9E43F03F)) {
```
Comment (rgrover) [inline source/sleep.c]: what is `0x9E43F03F`?

Comment (pan-) [inline source/sleep.c]: The mask for reserved bits of this register.

Comment (rgrover) [inline source/sleep.c]: could you please make that more explicit by using a symbolic constant?

Comment (pan-) [inline source/sleep.c]: Yes, I've just moved it to a `#define`.

**Comment by emidttun** [comment]
Looks good to me.

**Comment by emidttun** [comment]
@0xc0170: So the __disable_irq() only sets PRIMASK? Wasn't aware of it, but in that case I have no further comments.

File: source/sleep.c [hunk 2]
```diff
@@ -37,8 +41,31 @@ void mbed_enter_sleep(sleep_t *obj)
...
+        // interrupt, return from here. If no interrupts are pending, just call
+        // WFE.
+
+        // Set an event and wake up whatsoever, this will clear the event
+        // register from all previous events set (SVC call included)
+        __SEV();
         __WFE();
+
+        // Test if there is an interrupt pending (mask reserved regions)
+        if(SCB->ICSR & (SCB_ICSR_RESERVED_BITS_MASK)) {
```
Comment (rgrover) [inline source/sleep.c]: coding style: please add a space between `if` and `(`.

**Comment by 0xc0170** [comment]
LGTM

**Comment by bogdanm** [comment]
LGTM, +1
@0xc0170, not sure what style issue you're referring to. Please let me know if you think this is OK to merge.

**Comment by 0xc0170** [comment]
It was fixed in the commit above. I'll merge this

**Pull request merged**